### ISSUE TITLE
fix `SyntaxError: Unexpected token ] in JSON at position 0` in cache

### ISSUE THREAD

**Issue by ccasher** [cloud-carbon-footprint/cloud-carbon-footprint#1197]
### Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/CONTRIBUTING.md) for this project.
- [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details
For a while when we have had issues to the cache, we have suggested to just delete the file and re-run, but we should really investigate why this is a recurring issue.

```
SyntaxError: Unexpected token ] in JSON at position 0
    at JSON.parse (<anonymous>)
    at DelimitedStream.<anonymous> (/...node_modules/@cloud-carbon-footprint/app/dist/common/helpers.js:41:31)
    
```

### Expected Behavior

When making repeated calls to the cache file, it should continue to return estimates with no issues.

### Actual Behavior

Occasionally, after making subsequent calls to the cache, the following error occurs:
`SyntaxError: Unexpected token ] in JSON at position 0`

### To Reproduce

Trigger estimations, and try to refresh to pull from the cache multiple times. If using the backstage plugin, try to make the initial request to `/cloud-carbon-footprint`, then visit another route and return back to the ccf route.

© 2021 Thoughtworks, Inc.

**Comment by oconnormj** [comment]
Hi @4upz,

I'm experiencing the above described behaviour in a repeatable manner when using the [Backstage Plugin](https://www.cloudcarbonfootprint.org/docs/backstage-plugins) and deploying into Kubernetes. I understand that MongoDb is the preferred cache store for my scenario but I am only at the POC stage and I'm attempting to avoid the cost overhead of MongDB.

I'm really interested in the outcome of this bugfix, if any. Much appreciate the effort that you are putting in on this.

Do you have any idea of the timeline involved to provide a fix?

**Comment by 4upz** [comment]
Hi @oconnormj, sorry that this issue is blocking you. My work on this has slowed a bit due to the recent holiday but is currently starting to pick back up. I plan on having it addressed in time for our December release which we're aiming to do within the next few weeks.

Just a heads up, both [MongoDB Atlas ](https://www.mongodb.com/docs/atlas/tutorial/deploy-free-tier-cluster/)(Cloud) and [MongoDB Community](https://www.mongodb.com/try/download/community) (Local) both have free tiers that are more than suitable for POC and most CCF use cases. If cost is the only factor, I'd recommend looking into those options as it should be able to unblock you without the need for paying for the service, and the setup is pretty straightforward.

However, I also understand if you wish to forgo the database setup or have other company-related blockers that make MongoDB non-ideal for the POC stage. You can follow this issue for updates on continued progress.

**Comment by 4upz** [comment]
Seems that this issue occurs when parsing the cache file after adding two separate date ranges. When requesting non-inclusive date ranges. For example:

<ins>**Request 1**</ins>
start=2023-01-01
end=2021-01-02
**Result**: Request successful, cache file created and data is added

<ins>**Request 2**</ins>
start=2023-01-01
end=2021-01-04
**Result**: Request successful, cached data returned alongside added new data

<ins>**Request 3**</ins>
start=2023-02-01
end=2021-02-02
**Result**: Error occurs, and server becomes unresponsive due to it not being caught

The same goes for if you were to do unconnected dates in the same month, then the second date range would be successful but an error will be thrown the next time that the unconnected dates are read from the cache. Gonna investigate if this is a reading error, or an error with the way that we are saving separate date ranges to the cache.

**Comment by 4upz** [comment]
The above commit should add graceful error handling to the file parsing issue allowing for the api to continue running when it encounters the error and to still return fresh estimates. If it detects the file parsing error, it will ignore the cache and grab fresh estimates.

My plan is to still get to the root cause and fix the issue with unconnected estimates not saving properly. However, if that's too difficult, then we will have the cache manager delete the corrupted file and create a new one with fresh estimates -- and then look into ways to revamp the LocalCacheManager to be less error prone with better timestamp handling/sorting.

**Comment by 4upz** [comment]
It seems the root issue is that the use of [fs.promises.write](https://nodejs.org/api/fs.html#filehandlewritestring-position-encoding) and the way that we're writing new estimates to the cache is overwriting previous content in the cache file. New estimates are written over the top of the file, leaving the format to be messed up and causing the JSON error when attempting to parse the file upon the next load. 

When writing an inclusive range of merged estimates as shown in the [helpers file](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/ce97b4c13e18b2026fc698106a564c9996528c12/packages/app/src/common/helpers.ts#L4) (partially already cached, partially new) the error does not occur since the content is long enough to completely write over the cache file (unless a longer unrelated range has been stored previously). When doing a non-inclusive range, the new range is added to the top of the file and cuts off or replaces previous data at the top of the file. 

In short, if the written data is longer than the old data in the file, then no error occurs but old data is erased. When data to be written is shorter than old data in the file, then error occurs due to content being partially overwritten causing us to break the JSON structure and corrupting the cached data.

Solution should be to appropriately append new estimates to the file. This may even require reading the full cache then properly merging it. Care should be taken when doing this to make sure that previously cached data isn't affected by the newly written data while also not blindly appending and causing duplicate data.

**Comment by 4upz** [comment]
Fix is pushed and is pending review
